lib/gems/pending/util/object_storage/miq_swift_storage.rb

```diff
@@ -0,0 +1,148 @@
+    @bucket_name         = URI(@settings[:uri]).host
+
+    raise "username and password are required values!" if @settings[:username].nil? || @settings[:password].nil?
+    _scheme, _userinfo, @host, @port, _registry, @mount_path, _opaque, query, _fragment = URI.split(URI.encode(@settings[:uri]))
+    query_params(query)
+    @swift          = nil
+    @username       = @settings[:username]
+    @password       = @settings[:password]
+    @container_name = @mount_path[0] == File::Separator ? @mount_path[1..-1] : @mount_path
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "swift:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "swift://" and the container name from the URI"
+    # Also remove teh leading delimiter.
+    object_file_with_bucket = URI.split(URI.encode(remote_file))[5]
+    object_file_with_bucket.split(File::Separator)[2..-1].join(File::Separator)
+  end
+
+  def upload_single(dest_uri)
+    #
+    # Get the remote path, and parse out the bucket name.
+    #
+    object_file = uri_to_object_path(dest_uri)
+    #
+    # write dump file to swift
+    #
+    logger.debug("Writing [#{source_input}] to => Bucket [#{container_name}] using object file name [#{object_file}]")
+    begin
+      swift_file = container.files.new(:key => object_file)
+      params     = {
+        :expects       => [201, 202],
+        :headers       => {},
+        :request_block => -> { read_single_chunk },
+        :idempotent    => false,
+        :method        => "PUT",
+        :path          => "#{Fog::OpenStack.escape(swift_file.directory.key)}/#{Fog::OpenStack.escape(swift_file.key)}"
+      }
+      swift_file.service.send(:request, params)
+      clear_split_vars
+    rescue Excon::Errors::Unauthorized => err
+      logger.error("Access to Swift container #{@container_name} failed due to a bad username or password. #{err}")
+      msg = "Access to Swift container #{@container_name} failed due to a bad username or password. #{err}"
+      raise err, msg, err.backtrace
+    rescue => err
+      logger.error("Error uploading #{source_input} to Swift container #{@container_name}. #{err}")
+      msg = "Error uploading #{source_input} to Swift container #{@container_name}. #{err}"
+      raise err, msg, err.backtrace
+    end
+  end
+
+  def mkdir(_dir)
+    container
+  end
+
+  def container
+    @container ||= begin
+                     container   = swift.directories.get(container_name)
+                     logger.debug("Swift container [#{container}] found") if container
+                     container ||= create_container
+                     container
+                   rescue Fog::Storage::OpenStack::NotFound
+                     logger.debug("Swift container #{container_name} does not exist.  Creating.")
+                     create_container
```

Comment:
I don't understand this logic. Where is this error being raised? Is there one case where `swift.directories.get` returns nil and some other one where it raises?

Reply:
For what it is worth, I put a comment in my branch about this change in this commit message: https://github.com/ManageIQ/manageiq-gems-pending/commit/36fa87af3b4edc96ad31d48d026b767d4e774bb8

Reply:
Ah that's valuable info. I think that's comment-worthy actually.

This is fine as is then if we add a comment mentioning that there are some cases where that call fails in the two different ways.

Reply:
Done.